app/desktop/studio_server/tools_api.py

```diff
@@ -0,0 +1,68 @@
+from typing import Any, Dict, List
+
+from fastapi import FastAPI, HTTPException
+from kiln_ai.datamodel.basemodel import ID_TYPE
+from kiln_ai.datamodel.external_tool import ExternalTool, ToolType
+from kiln_server.project_api import project_from_id
+from pydantic import BaseModel, ValidationError
+
+
+class KilnToolDescription(BaseModel):
+    name: str
+    id: ID_TYPE
+    description: str | None
+
+
+class ExternalToolCreationRequest(BaseModel):
+    name: str
+    server_url: str
+    headers: Dict[str, Any] = {}
+    description: str | None = None
+
+
+def connect_tools_api(app: FastAPI):
+    @app.get("/api/projects/{project_id}/available_tools")
+    async def get_available_tools(
+        project_id: str,
+    ) -> List[KilnToolDescription]:
+        project = project_from_id(project_id)
+
+        return [
+            KilnToolDescription(
+                name=tool.name,
+                id=tool.id,
+                description=tool.description,
+            )
+            for tool in project.external_tools()
+        ]
+
+    @app.post("/api/projects/{project_id}/connect_external_tool")
+    async def connect_external_tool(
+        project_id: str, tool_data: ExternalToolCreationRequest
```

Comment:
This is fine, but for future reference you can also just do a list of properties here if you don’t need the BaseModel